S3TransferUtilitySample/src/com/amazonaws/demo/s3transferutility/UploadActivity.java

```diff
@@ -519,85 +517,68 @@ private void beginUploadInBackground(String filePath) {
     }
 
     /*
-     * Gets the file path of the given Uri.
+     * Copies the resource associated with the Uri to a new File in the cache directory.
+     * @param uri The Uri.
+     * @return a new File representing the Uri in the cache directory.
+
      */
-    @SuppressLint("NewApi")
-    private String getPath(Uri uri) throws URISyntaxException {
-        final boolean needToCheckUri = Build.VERSION.SDK_INT >= 19;
-        String selection = null;
-        String[] selectionArgs = null;
-        // Uri is different in versions after KITKAT (Android 4.4), we need to
-        // deal with different Uris.
-        if (needToCheckUri && DocumentsContract.isDocumentUri(getApplicationContext(), uri)) {
-            if (isExternalStorageDocument(uri)) {
-                final String docId = DocumentsContract.getDocumentId(uri);
-                final String[] split = docId.split(":");
-                return Environment.getExternalStorageDirectory() + "/" + split[1];
-            } else if (isDownloadsDocument(uri)) {
-                final String id = DocumentsContract.getDocumentId(uri);
-                uri = ContentUris.withAppendedId(
-                        Uri.parse("content://downloads/public_downloads"), Long.valueOf(id));
-            } else if (isMediaDocument(uri)) {
-                final String docId = DocumentsContract.getDocumentId(uri);
-                final String[] split = docId.split(":");
-                final String type = split[0];
-                if ("image".equals(type)) {
-                    uri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
-                } else if ("video".equals(type)) {
-                    uri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
-                } else if ("audio".equals(type)) {
-                    uri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
+    private File getFile(Uri uri) throws FileNotFoundException {
+        final InputStream in = getContentResolver().openInputStream(uri);
+        OutputStream out = null;
+        try {
+            final File file = new File(getCacheDir(), getFileName(uri));
+            out = new FileOutputStream(file, false);
+            byte[] buffer = new byte[1024];
+            for (int len; (len = in.read(buffer)) != -1; ) {
+                out.write(buffer, 0, len);
+            }
+            return file;
+        } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
At minimum, this should move to `Log.w(TAG, "Failed to read file or whatever...", fileReadFailure);`, where-in:
1. exception variable has a semantic name
2. `android.util.Log` i used in place of `printStackTrace()` 

Alternately, you could just allow the exception to bubble up, by adding it to the cascade of method signatures.

Alternately to both, you could wrap this in a `new RuntimeException(...)`, and then not have to add it to any method signatures.

If it were me, and for the purposes of this sample application code -- I'd probably do that last option, to reduce extra boiler-plate, but still ensure a valid output of this utility function (or else, a well known crash.)

Reply:
__UPDATE__: Sorry, the pattern in the other error states is to invoke `Toast.makeText(...)`. That could work well here, too, instead of crashing / logging to console, etc.

Reply:
I've updated this method to throw the IOException, and show a Toast when catching it downstream.